File: backend/LimsAuth.Api/Models/Schedule/TeachingApplication.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace LimsAuth.Api.Models;

[Table("teaching_applications")]
public class TeachingApplication
{
    [Key]
    [Column("id")]
    public Guid Id { get; set; } = Guid.NewGuid();

    [Required]
    [Column("semester_id")]
    public Guid SemesterId { get; set; }

    [Required]
    [Column("teaching_task_id")]
    public Guid TeachingTaskId { get; set; }

    [Required]
    [Column("course_name")]
    [MaxLength(200)]
    public string CourseName { get; set; } = string.Empty;

    [Required]
    [Column("major_id")]
    public Guid MajorId { get; set; }

    [Required]
    [Column("major_name")]
    [MaxLength(100)]
    public string MajorName { get; set; } = string.Empty;

    [Required]
    [Column("class_id")]
    public Guid ClassId { get; set; }

    [Required]
    [Column("class_name")]
    [MaxLength(100)]
    public string ClassName { get; set; } = string.Empty;

    [Column("week_numbers")]
    [MaxLength(200)]
    public string? WeekNumbersJson { get; set; }

    [NotMapped]
    public List<int> WeekNumbers
    {
        get => string.IsNullOrEmpty(WeekNumbersJson)
            ? new List<int>()
            : System.Text.Json.JsonSerializer.Deserialize<List<int>>(WeekNumbersJson) ?? new List<int>();
        set => WeekNumbersJson = System.Text.Json.JsonSerializer.Serialize(value);
    }

    [Column("day_of_week")]
    public int DayOfWeek { get; set; }

    [Column("period_numbers")]
    [MaxLength(200)]
    public string? PeriodNumbersJson { get; set; }

    [NotMapped]
    public List<int> PeriodNumbers
    {
        get => string.IsNullOrEmpty(PeriodNumbersJson)
            ? new List<int>()
            : System.Text.Json.JsonSerializer.Deserialize<List<int>>(PeriodNumbersJson) ?? new List<int>();
        set => PeriodNumbersJson = System.Text.Json.JsonSerializer.Serialize(value);
    }

    [Column("expected_lab_id")]
    public Guid? ExpectedLabId { get; set; }

    [Column("remark")]
    [MaxLength(500)]
    public string? Remark { get; set; }

    [Required]
    [Column("applicant_id")]
    public Guid ApplicantId { get; set; }

    [Required]
    [Column("applicant_name")]
    [MaxLength(100)]
    public string ApplicantName { get; set; } = string.Empty;

    [Column("status")]
    [MaxLength(20)]
    public ApprovalStatus Status { get; set; } = ApprovalStatus.Pending;

    [Column("approval_comment")]
    [MaxLength(500)]
    public string? ApprovalComment { get; set; }

    [Column("approved_by")]
    public Guid? ApprovedBy { get; set; }

    [Column("approved_at")]
    public DateTime? ApprovedAt { get; set; }

    [Column("is_cancelled")]
    public bool IsCancelled { get; set; }

    [Column("cancel_reason")]
    [MaxLength(500)]
    public string? CancelReason { get; set; }

    [Column("created_at")]
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

    [Column("created_by")]
    public string? CreatedBy { get; set; }

    [Column("updated_at")]
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;

    [Column("updated_by")]
    public string? UpdatedBy { get; set; }

    public virtual Semester? Semester { get; set; }
    public virtual TeachingTask? TeachingTask { get; set; }
    public virtual Major? Major { get; set; }
    public virtual Class? Class { get; set; }
    public virtual Lab? ExpectedLab { get; set; }
    public virtual User? Applicant { get; set; }
}

File: mobile/lib/models/teaching_application.dart
import 'lab.dart';
import '../utils/num_utils.dart';

class TeachingApplication {
  final String id;
  final String? semesterId;
  final String? semesterName;
  final String? teachingTaskId;
  final String? courseName;
  final String? majorId;
  final String? majorName;
  final String? classId;
  final String? className;
  final List<int>? weekNumbers;
  final int dayOfWeek;
  final List<int>? periodNumbers;
  final String? expectedLabId;
  final String? expectedLabName;
  final String? remark;
  final String? applicantId;
  final String? applicantName;
  final String? status;
  final String? approvalComment;
  final String? approvedBy;
  final String? approverName;
  final String? approvedAt;
  final bool isCancelled;
  final String? cancelReason;
  final String? createdAt;

  TeachingApplication({
    required this.id,
    this.semesterId,
    this.semesterName,
    this.teachingTaskId,
    this.courseName,
    this.majorId,
    this.majorName,
    this.classId,
    this.className,
    this.weekNumbers,
    this.dayOfWeek = 0,
    this.periodNumbers,
    this.expectedLabId,
    this.expectedLabName,
    this.remark,
    this.applicantId,
    this.applicantName,
    this.status,
    this.approvalComment,
    this.approvedBy,
    this.approverName,
    this.approvedAt,
    this.isCancelled = false,
    this.cancelReason,
    this.createdAt,
  });

  factory TeachingApplication.fromJson(Map<String, dynamic> json) {
    return TeachingApplication(
      id: json['id']?.toString() ?? '',
      semesterId: json['semesterId']?.toString(),
      semesterName: json['semesterName']?.toString(),
      teachingTaskId: json['teachingTaskId']?.toString(),
      courseName: json['courseName']?.toString(),
      majorId: json['majorId']?.toString(),
      majorName: json['majorName']?.toString(),
      classId: json['classId']?.toString(),
      className: json['className']?.toString(),
      weekNumbers: (json['weekNumbers'] as List<dynamic>?)?.cast<int>(),
      dayOfWeek: safeInt(json['dayOfWeek']),
      periodNumbers: (json['periodNumbers'] as List<dynamic>?)?.cast<int>(),
      expectedLabId: json['expectedLabId']?.toString(),
      expectedLabName: json['expectedLabName']?.toString(),
      remark: json['remark']?.toString(),
      applicantId: json['applicantId']?.toString(),
      applicantName: json['applicantName']?.toString(),
      status: json['status']?.toString(),
      approvalComment: json['approvalComment']?.toString(),
      approvedBy: json['approvedBy']?.toString(),
      approverName: json['approverName']?.toString(),
      approvedAt: json['approvedAt']?.toString(),
      isCancelled: json['isCancelled'] as bool? ?? false,
      cancelReason: json['cancelReason']?.toString(),
      createdAt: json['createdAt']?.toString(),
    );
  }

  Map<String, dynamic> toJson() => {
        'semesterId': semesterId,
        'teachingTaskId': teachingTaskId,
        'courseName': courseName,
        'majorId': majorId,
        'majorName': majorName,
        'classId': classId,
        'className': className,
        'weekNumbers': weekNumbers,
        'dayOfWeek': dayOfWeek,
        'periodNumbers': periodNumbers,
        'expectedLabId': expectedLabId,
        'remark': remark,
      };
}

class CreateTeachingApplicationRequest {
  final String semesterId;
  final String teachingTaskId;
  final String? courseName;
  final String? majorId;
  final String? majorName;
  final String? classId;
  final String? className;
  final List<int>? weekNumbers;
  final int dayOfWeek;
  final List<int>? periodNumbers;
  final String? expectedLabId;
  final String? remark;

  CreateTeachingApplicationRequest({
    required this.semesterId,
    required this.teachingTaskId,
    this.courseName,
    this.majorId,
    this.majorName,
    this.classId,
    this.className,
    this.weekNumbers,
    this.dayOfWeek = 0,
    this.periodNumbers,
    this.expectedLabId,
    this.remark,
  });

  Map<String, dynamic> toJson() => {
        'semesterId': semesterId,
        'teachingTaskId': teachingTaskId,
        'courseName': courseName,
        'majorId': majorId,
        'majorName': majorName,
        'classId': classId,
        'className': className,
        'weekNumbers': weekNumbers,
        'dayOfWeek': dayOfWeek,
        'periodNumbers': periodNumbers,
        'expectedLabId': expectedLabId,
        'remark': remark,
      };
}
